File: backend/app/schemas/monetization.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict


# ===================================================================
# PLANS
# ===================================================================
class PlanBase(BaseModel):
    name: str
    description: str | None = None
    price: float
    currency: str = "INR"
    interval: str = "monthly"
    features: list[str] | None = None
    is_active: bool = True


class PlanCreate(PlanBase):
    pass


class PlanUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    price: float | None = None
    currency: str | None = None
    interval: str | None = None
    features: list[str] | None = None
    is_active: bool | None = None


class PlanResponse(PlanBase):
    id: int
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ===================================================================
# COUPONS
# ===================================================================
class CouponBase(BaseModel):
    code: str
    discount_type: str = "percentage"  # percentage, fixed
    discount_amount: float
    expiry_date: datetime | None = None
    is_active: bool = True


class CouponCreate(CouponBase):
    pass


class CouponResponse(CouponBase):
    id: int
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class CouponValidateRequest(BaseModel):
    code: str
    plan_id: int


# ===================================================================
# SUBSCRIPTIONS
# ===================================================================
class SubscriptionBase(BaseModel):
    plan_id: int
    status: str = "active"
    current_period_start: datetime
    current_period_end: datetime
    cancel_at_period_end: bool = False
    razorpay_subscription_id: str | None = None
    coupon_id: int | None = None


class SubscriptionCreate(SubscriptionBase):
    user_id: int


class SubscriptionResponse(SubscriptionBase):
    id: int
    created_at: datetime
    updated_at: datetime
    plan: PlanResponse | None = None
    coupon: CouponResponse | None = None

    model_config = ConfigDict(from_attributes=True)


# ===================================================================
# PAYMENTS
# ===================================================================
class PaymentBase(BaseModel):
    subscription_id: int | None = None
    amount: float
    currency: str = "INR"
    status: str = "success"
    provider: str = "razorpay"
    transaction_id: str


class PaymentCreate(PaymentBase):
    user_id: int


class PaymentResponse(PaymentBase):
    id: int
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ===================================================================
# INVOICES
# ===================================================================
class InvoiceResponse(BaseModel):
    id: int
    user_id: int
    subscription_id: int
    billing_name: str
    billing_email: str
    amount: float
    currency: str
    status: str
    pdf_url: str | None = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ===================================================================
# TRANSACTIONS
# ===================================================================
class TransactionResponse(BaseModel):
    id: int
    user_id: int
    amount: float
    type: str
    description: str | None = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ===================================================================
# CHECKOUT & VERIFICATION SESSIONS
# ===================================================================
class CheckoutSessionRequest(BaseModel):
    plan_id: int
    coupon_code: str | None = None


class CheckoutSessionResponse(BaseModel):
    payment_order_id: str
    amount: float
    currency: str
    checkout_url: str | None = None
    active_provider: str
    is_sandbox: bool = False


class PaymentVerificationRequest(BaseModel):
    razorpay_payment_id: str
    razorpay_order_id: str
    razorpay_signature: str
    plan_id: int
    coupon_code: str | None = None


# ===================================================================
# ADMIN STATS
# ===================================================================
class MonthlyRevenuePoint(BaseModel):
    month: str
    revenue: float


class AdminRevenueStats(BaseModel):
    total_revenue: float
    mrr: float
    active_subscribers: int
    monthly_data: list[MonthlyRevenuePoint]
    active_subscribers_by_plan: dict[str, int]

File: frontend/src/types/monetization.ts
export interface PlanResponse {
  id: number;
  name: string;
  description: string | null;
  price: number;
  currency: string;
  interval: string;
  features: string[] | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface CouponResponse {
  id: number;
  code: string;
  discount_type: string;
  discount_amount: number;
  expiry_date: string | null;
  is_active: boolean;
  created_at: string;
}

export interface SubscriptionResponse {
  id: number;
  user_id: number;
  plan_id: number;
  status: string;
  current_period_start: string;
  current_period_end: string;
  cancel_at_period_end: boolean;
  razorpay_subscription_id: string | null;
  coupon_id: number | null;
  created_at: string;
  updated_at: string;
  plan?: PlanResponse;
  coupon?: CouponResponse;
}

export interface InvoiceResponse {
  id: number;
  user_id: number;
  subscription_id: number;
  billing_name: string;
  billing_email: string;
  amount: number;
  currency: string;
  status: string;
  pdf_url: string | null;
  created_at: string;
}

export interface TransactionResponse {
  id: number;
  user_id: number;
  amount: number;
  type: string;
  description: string | null;
  created_at: string;
}

export interface CheckoutSessionResponse {
  payment_order_id: string;
  amount: number;
  currency: string;
  checkout_url: string | null;
  active_provider: string;
  is_sandbox?: boolean;
}

export interface MonthlyRevenuePoint {
  month: string;
  revenue: number;
}

export interface AdminRevenueStats {
  total_revenue: number;
  mrr: number;
  active_subscribers: number;
  monthly_data: MonthlyRevenuePoint[];
  active_subscribers_by_plan: Record<string, number>;
}
